retransmit.go

```diff
@@ -0,0 +1,108 @@
+package coap
+
+import (
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+)
+
+const ACK_TIMEOUT = 2000
+
+const ACK_MAX_TIMEOUT = 3000
+
+const MAX_RETRANSMIT = 4
+
+// generate a random timeout following the RFC
+func randTimeout() time.Duration {
+	return time.Duration((rand.Intn(ACK_MAX_TIMEOUT-ACK_TIMEOUT) + ACK_TIMEOUT)) * time.Millisecond
+}
+
+// a CoAP retransmitter
+type Retransmitter struct {
+	inflight map[string]*flight // currently non-acked CON messages
+	lock     sync.RWMutex
+	buf      []byte       // receive buffer
+	c        *net.UDPConn // UDP socket
+
+}
+
+// a CoAP Confirmable message in-flight
+type flight struct {
+	msg     *Message      // the message to be retransmit if needed
+	ack     chan struct{} // channel for signaling ack is received
+	retrans int           // number of retransmission tentatives
+}
+
+func NewRetransmitter(c *net.UDPConn) *Retransmitter {
+	s := new(Retransmitter)
+	s.c = c
+	s.inflight = make(map[string]*flight)
+	s.buf = make([]byte, 1200)
+	return s
+}
+
+func (retrans *Retransmitter) Record(req Message, a *net.UDPAddr) error {
```

Comment:
do we want to pass in some channel or something so that we can keep track of messages that were never delivered? Or, do we maybe want to provide the option to block until received for clients?